### DISCUSSION TITLE
Reconsider deprecating Object as a GraphQL context

### DISCUSSION THREAD

**Discussion by vojtapol** [graphql-java/graphql-java#2515] [Ideas]
https://github.com/graphql-java-kickstart/graphql-java-tools/pull/566

This is an  unfortunate PR. The context object has been used by many people in many ways. I don't think the maintainers considered them all. Lot of the usage of the context object had nothing to do with just storing data. For example we have put behavior methods in the context and it has been working for us flawlessly.

Even for plain data storage the new class is a step back. Instead of something like `context.getUser()` we will now have to do `context.get("user")` with possibility for typos and no type safety.

I wish this was first discussed with the community.

## Proposal
1. Keep `getGraphqlContext()` but keep `getContext()` forever without deprecating it, OR
2. Make `GraphQLContext` extendable by anyone, because why not?

**Comment by tsegismont** [comment]
In the PR that introduced `GraphQLContext` it is [acknowledged](https://github.com/graphql-java/graphql-java/pull/2368#issue-660621714) that the legacy `getContext` is key to any GraphQL-Java application:

>At some point the legacy getContext will disappear but since this is a WIDELY used mechanism that will not be any time soon.

That's not _forever_ of course but that sounds reasonable to me.

Regarding your concern about type safety, the `GraphQLContext` methods take any type of object as key, so you don't have to use a `String`. You could do something like `graphQLContext.get(User.class)` (just an example).

As a library maintainer, I believe the change is a move in the right direction for the reasons explained in https://github.com/graphql-java/graphql-java/issues/2357#issue-896681925.

**Reply by vojtapol** [reply]
> That's not forever of course but that sounds reasonable to me.

As you can see below the plan seems to be to remove it in the next two versions. That does not sound reasonable for a WIDELY used mechanism without providing a new on-par solution.

> You could do something like graphQLContext.get(User.class) (just an example).

Using `Class` instead of `String` for a key gives me no type safety about the **value** I am getting back.

**Reply by tsegismont** [reply]
Right, no type safety about the value, but that is not different from using `context()` which returns an `Object` too.

**Comment by bbakerman** [comment]
We added the GraphQLContext signature expressly for frameworks like Spring or graphql-java-tools.

After working with the developers from Spring, they convinced is that a context object that could contain "object named" things would work better than the single opaque object we had.


The problem with the opaque object was that it was TOO opaque.  There was no way for 2 bits of code (user code vs framework ode say) to know the shape of the context and hence safely put context data away.     Now developers / frameworks can.


This would allow spring code to put say `graphqlContext.put(org.spring.SpringCode.class, someSpringSpecificContextData)` into  a named place which would NOT interfere with some other bit of code from DGS or graphql-java-tools say.

graphql-java-tools could for example choose to have their own `JavaToolsContext.class` object say and wrap and unwrap it before giving it out in its layers of code OR it could choose to expose the `GraphqlContext` direct say.

At some time, probably in 2 versions, we will remove the opaque context object (which by the way defaulted to a `GraphqlContext` object but you would not know it via signature) .

**Comment by bbakerman** [comment]
> Make GraphQLContext extendable by anyone, because why not?

It is extendable - its a map that you can put stuff into by name - a big old map of STUFF.  You can put as much well typed objects into it by key and get them out.  

You don't need to class extend it  - you need to decide how to put your part of the code world into it.

 (inheritance vs composition / you need to compose with it)

**Comment by vojtapol** [comment]
> It is extendable - its a map that you can put stuff into by name - a big old map of STUFF. You can put as much well typed objects into it by key and get them out.
>
> You don't need to class extend it - you need to decide how to put your part of the code world into it.

The insertion and extraction of objects into the map is not type safe in any way. You have to cast it on extraction. If the value stored under the key has unexpected type you will get a run-time exception.

> You don't need to class extend it - you need to decide how to put your part of the code world into it.

Our GraphQL context (not talking about `tools` here) contains both state and behavior. Storing behavior as data is somewhat possible I guess but very cumbersome.

Can you please elaborate on why normal inheritance extension is not being allowed? I understand you are trying to cater to libraries but I am speaking here from the experience of a regular user. We **need** to add our own fully typed methods and behavior to `GraphQLContext`.

The only workaround I can think of is to add just one item into the new `GraphQLContext` like so:

```java
graphqlContext.put("my-context", myContext);
```
and then use it like so:
```java
MyContext context = graphqlContext.get("my-context");
context.callMyTypedMethod();
```

Is that hacky or is this what you would recommend?

**Comment by bbakerman** [comment]
I am not sure how much graphql-java-tools wants hide the raw DataFetchingEnvironment and whether it provides its own wrapper ?

Basically if you have "behavior" stored in context that you would like "passed" back to you then JUST put the behavior in a place that is not likely to clash with any other.   That's why "classes as keys" a  good pattern.

```
    ToolsBehavior behavior = context.get(ToolsBehavior.class)
```

The challenge we have (and graphql-java-tools has and graphql-java-spring has and Netflix DGS has...) is that if you are the only consumer of a context object then you can have one well known shape - your so called type safe mechanism.  However you have proven your not the only consumer because you have context with graphql-java-tools behavior in it and also need it to be a general context place for consumers of your framework so they can put stuff away (unless you dont allow that in which case - thats strange - how do people do custom stuff??)

As for casting "type safe" objects - you already have to to a cast - the signature is `Object context` - so today you need to do

```
    ToolsContext behavior = (ToolsContext) context;
```

If it's not what you think you will get a class cast exception.  With the `graphlContext.get(someKey)` method you will geta  class cast exception if its not what you think.   See its the same thing - no better no worse.

Try this thought experiement - imagine some one build a framework on top of your framework - graphql-better-tools - and they wanted to put away "better-tools" context along side the base framework context.... well how do you prevent clashes?  How can they their own "specific context" while allowing the underlying framework to retain its specific context?

It can't be via inheritance because ToolsContext and BetterToolsContext cant be known to be compatible.

But the map of stuff, which is really a namespaced place of context, allows you to store away your context and the other code to store away its context.

**Comment by vojtapol** [comment]
@bbakerman Your explanation makes sense. Now I understand the reasoning behind this change and how to utilize it. Thank you for writing this up.

As it stands I see no reason to not proceed with the deprecation and move to this new mechanism. 👍
